File: typecast-dart/lib/src/timestamps.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'models.dart';

class AlignmentSegmentWord {
  const AlignmentSegmentWord({
    required this.word,
    required this.startTime,
    required this.endTime,
  });

  factory AlignmentSegmentWord.fromJson(Map<String, dynamic> json) =>
      AlignmentSegmentWord(
        word: json['word'] as String? ?? '',
        startTime: (json['start_time'] as num?)?.toDouble() ?? 0,
        endTime: (json['end_time'] as num?)?.toDouble() ?? 0,
      );

  final String word;
  final double startTime;
  final double endTime;
}

class AlignmentSegmentCharacter {
  const AlignmentSegmentCharacter({
    required this.character,
    required this.startTime,
    required this.endTime,
  });

  factory AlignmentSegmentCharacter.fromJson(Map<String, dynamic> json) =>
      AlignmentSegmentCharacter(
        character: json['character'] as String? ?? '',
        startTime: (json['start_time'] as num?)?.toDouble() ?? 0,
        endTime: (json['end_time'] as num?)?.toDouble() ?? 0,
      );

  final String character;
  final double startTime;
  final double endTime;
}

class TtsWithTimestampsResponse {
  const TtsWithTimestampsResponse({
    required this.audio,
    required this.audioFormat,
    required this.audioDuration,
    this.words = const [],
    this.characters = const [],
  });

  factory TtsWithTimestampsResponse.fromJson(Map<String, dynamic> json) =>
      TtsWithTimestampsResponse(
        audio: json['audio'] as String? ?? '',
        audioFormat:
            json['audio_format'] == 'mp3' ? AudioFormat.mp3 : AudioFormat.wav,
        audioDuration: (json['audio_duration'] as num?)?.toDouble() ?? 0,
        words: (json['words'] as List? ?? [])
            .map((item) => AlignmentSegmentWord.fromJson((item as Map).cast()))
            .toList(),
        characters: (json['characters'] as List? ?? [])
            .map(
              (item) =>
                  AlignmentSegmentCharacter.fromJson((item as Map).cast()),
            )
            .toList(),
      );

  final String audio;
  final AudioFormat audioFormat;
  final double audioDuration;
  final List<AlignmentSegmentWord> words;
  final List<AlignmentSegmentCharacter> characters;

  Uint8List audioBytes() => base64Decode(audio);

  Future<void> saveAudio(String path) async {
    await File(path).writeAsBytes(audioBytes());
  }

  String toSrt() => _formatCaptions(webVtt: false);

  String toVtt() => 'WEBVTT\n\n${_formatCaptions(webVtt: true)}';

  String _formatCaptions({required bool webVtt}) {
    if (words.isEmpty && characters.isEmpty) {
      throw StateError('No alignment segments are available');
    }
    final segments = words.isNotEmpty
        ? words.map(
            (word) => _CaptionSegment(word.word, word.startTime, word.endTime),
          )
        : characters.map(
            (char) =>
                _CaptionSegment(char.character, char.startTime, char.endTime),
          );
    final lines = <String>[];
    var index = 1;
    for (final segment in segments) {
      if (!webVtt) lines.add('${index++}');
      lines.add(
        '${_formatTimestamp(segment.start, webVtt)} --> '
        '${_formatTimestamp(segment.end, webVtt)}',
      );
      lines
        ..add(segment.text)
        ..add('');
    }
    return lines.join('\n');
  }
}

class _CaptionSegment {
  const _CaptionSegment(this.text, this.start, this.end);

  final String text;
  final double start;
  final double end;
}

String _formatTimestamp(double seconds, bool webVtt) {
  final millis = (seconds * 1000).round();
  final hours = millis ~/ 3600000;
  final minutes = (millis % 3600000) ~/ 60000;
  final secs = (millis % 60000) ~/ 1000;
  final ms = millis % 1000;
  final decimal = webVtt ? '.' : ',';
  return '${hours.toString().padLeft(2, '0')}:'
      '${minutes.toString().padLeft(2, '0')}:'
      '${secs.toString().padLeft(2, '0')}$decimal'
      '${ms.toString().padLeft(3, '0')}';
}

File: typecast-ruby/lib/typecast/timestamps.rb
require "base64"

module Typecast
  module Models
    class AlignmentSegmentWord
      attr_reader :word, :start_time, :end_time

      def self.from_h(hash)
        new(word: hash.fetch("word", ""), start_time: hash.fetch("start_time", 0).to_f, end_time: hash.fetch("end_time", 0).to_f)
      end

      def initialize(word:, start_time:, end_time:)
        @word = word
        @start_time = start_time
        @end_time = end_time
      end
    end

    class AlignmentSegmentCharacter
      attr_reader :character, :start_time, :end_time

      def self.from_h(hash)
        new(character: hash.fetch("character", ""), start_time: hash.fetch("start_time", 0).to_f, end_time: hash.fetch("end_time", 0).to_f)
      end

      def initialize(character:, start_time:, end_time:)
        @character = character
        @start_time = start_time
        @end_time = end_time
      end
    end

    class TTSWithTimestampsResponse
      attr_reader :audio, :audio_format, :audio_duration, :words, :characters

      def self.from_h(hash)
        new(
          audio: hash.fetch("audio", ""),
          audio_format: hash.fetch("audio_format", AUDIO_WAV),
          audio_duration: hash.fetch("audio_duration", 0).to_f,
          words: hash.fetch("words", []).map { |item| AlignmentSegmentWord.from_h(item) },
          characters: hash.fetch("characters", []).map { |item| AlignmentSegmentCharacter.from_h(item) }
        )
      end

      def initialize(audio:, audio_format:, audio_duration:, words: [], characters: [])
        @audio = audio
        @audio_format = audio_format
        @audio_duration = audio_duration
        @words = words
        @characters = characters
      end

      def audio_bytes
        Base64.decode64(audio)
      end

      def save_audio(path)
        File.binwrite(path, audio_bytes)
      end

      def to_srt
        caption_lines(webvtt: false)
      end

      def to_vtt
        "WEBVTT\n\n#{caption_lines(webvtt: true)}"
      end

      private

      def caption_lines(webvtt:)
        segments = words.empty? ? characters.map { |c| [c.character, c.start_time, c.end_time] } : words.map { |w| [w.word, w.start_time, w.end_time] }
        raise ArgumentError, "No alignment segments are available" if segments.empty?

        lines = []
        segments.each_with_index do |segment, index|
          lines << (index + 1).to_s unless webvtt
          lines << "#{format_time(segment[1], webvtt)} --> #{format_time(segment[2], webvtt)}"
          lines << segment[0]
          lines << ""
        end
        lines.join("\n")
      end

      def format_time(seconds, webvtt)
        millis = (seconds * 1000).round
        hours = millis / 3_600_000
        minutes = (millis % 3_600_000) / 60_000
        secs = (millis % 60_000) / 1000
        ms = millis % 1000
        format("%02d:%02d:%02d%s%03d", hours, minutes, secs, webvtt ? "." : ",", ms)
      end
    end
  end
end
